### PR TITLE
Fixed IEnumerable dense vector support

### PR DESCRIPTION
**Pull request by eisber**

### CODE REVIEW

**Comment by JohnLangford** [comment]
@lhoang29 can you take a look?

File: cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt [hunk 1]
```diff
@@ -155,6 +155,90 @@ namespace VW.Serializer
...
+            // support anchor feature
+            if (feature.AddAnchor)
+            {
+                context.NamespaceBuilder.AddFeature(ns.NamespaceHash, 1);
+                i++;
+            }
+
+            foreach (var v in value)
+            {
+				<# if(new[]{ typeof(Int64), typeof(UInt64), typeof(double) }.Contains(t)) {#>
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt]: maybe better to initialize a hash set of these types outside of the loop

File: cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt [hunk 2]
```diff
@@ -155,6 +155,90 @@ namespace VW.Serializer
...
+                context.NamespaceBuilder.AddFeature((ulong)(ns.NamespaceHash + i), (float)v);
+                i++;
+            }
+
+            if (context.StringExample == null)
+            {
+                return;
+            }
+
+			string featureString;
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt]: I've seen inconsistent whitespace issues with this file before, is it possible to fix?

File: cs/unittest/TestExpansion.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    [TestClass]
+    public class TestExpansionClass
+    {
+        [TestMethod]
+        [TestCategory("Marshal")]
+        public void TestExpansion()
+        {
+            using (var vw = new VowpalWabbitExampleValidator<ExpansionContext>(string.Empty))
+            {
+                vw.Validate("| 3:.1 4:.2 5:.3", new ExpansionContext() { Features = new[] { .1f, .2f, .3f }, Offset = 3 });
```
Comment (lhoang29) [inline cs/unittest/TestExpansion.cs]: thanks for adding the expansion support

File: cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt (hunk 2 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt]: I replace \t with 4x white space

File: cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt (hunk 1 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt]: this is done at build time. for readability I'd stick with the contains.

Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitDefaultMarshallerExt.tt]: Oh I see, these are compile-time directives

**Comment by lhoang29** [comment]
looks good now

**Pull request merged**